vendors/renesas/boards/rx65n-rsk/ports/ota_pal_for_aws/ota_pal.c

```diff
@@ -0,0 +1,1291 @@
+							tmp->content.length = fragmented_length;
+						}
+						else
+						{
+							assembled_length += tmp->content.length;
+							flash_block_candidate[loop_counter] = tmp;
+						}
+					}
+					else
+					{
+						break;
+					}
+					tmp = tmp->next;
+					loop_counter++;
+				}
+				else
+				{
+					break;
+				}
+			}
+		}
+
+		/* break if found completed flash_block_candidate or found end of list */
+		if((assembled_length == FLASH_CF_MIN_PGM_SIZE) || (tmp == NULL))
+		{
+			break;
+		}
+		/* search next candidate */
+		else
+		{
+			tmp = tmp->next;
+		}
+	}
+
+	/* assemble flash_block */
+	if(assembled_length == FLASH_CF_MIN_PGM_SIZE)
+	{
+		tmp = head;
+		/* remove flash_block from list */
+		flash_block->offset = flash_block_candidate[0]->content.offset;
+		flash_block->length = assembled_length;
+		for(uint32_t i = 0; i < loop_counter; i++)
+		{
+			memcpy(&flash_block->binary[index], flash_block_candidate[i]->content.binary, flash_block_candidate[i]->content.length);
+			index = flash_block_candidate[i]->content.length;
+			tmp = fragmented_flash_block_list_delete(tmp, flash_block_candidate[i]->content.offset);
+		}
+	}
+	else
+	{
+		tmp = head;
+		flash_block->offset = NULL;
+		flash_block->binary = NULL;
+		flash_block->length = NULL;
+	}
+	return tmp;
+}
+
+static void ota_flashing_task( void * pvParameters )
+{
+	flash_err_t flash_err;
+	static uint8_t block[(1 << otaconfigLOG2_FILE_BLOCK_SIZE)];
+	static uint32_t ulOffset;
+	static uint32_t length;
+
+	while(1)
+	{
+		xQueueReceive(xQueue, &packet_block_for_queue2, portMAX_DELAY);
+		xSemaphoreTake(xSemaphoreFlashig, portMAX_DELAY);
+		memcpy(block, packet_block_for_queue2.p_packet, packet_block_for_queue2.length);
+		ulOffset = packet_block_for_queue2.ulOffset;
+		length = packet_block_for_queue2.length;
+		flash_err = R_FLASH_Write((uint32_t)block, ulOffset + BOOT_LOADER_UPDATE_TEMPORARY_AREA_LOW_ADDRESS + BOOT_LOADER_USER_FIRMWARE_HEADER_LENGTH, length);
+		if(packet_block_for_queue2.length != 1024)
+		{
+			nop();
+		}
+		if(flash_err != FLASH_SUCCESS)
+		{
+			nop();
+		}
+		load_firmware_control_block.total_image_length += length;
+		vPortFree(packet_block_for_queue2.p_packet);
+	}
+
+}
+
+static void ota_flashing_callback(void *event)
+{
+	uint32_t event_code;
+	event_code = *((uint32_t*)event);
+
+    if((event_code != FLASH_INT_EVENT_WRITE_COMPLETE) || (event_code == FLASH_INT_EVENT_ERASE_COMPLETE))
+    {
+    	nop(); /* trap */
+    }
+	static portBASE_TYPE xHigherPriorityTaskWoken;
+	xSemaphoreGiveFromISR(xSemaphoreFlashig, &xHigherPriorityTaskWoken);
+}
+
+static void ota_header_flashing_callback(void *event)
+{
+	uint32_t event_code;
+	event_code = *((uint32_t*)event);
+
+	gs_header_flashing_task = OTA_FLASHING_COMPLETE;
+
+    if((event_code != FLASH_INT_EVENT_WRITE_COMPLETE) || (event_code == FLASH_INT_EVENT_ERASE_COMPLETE))
+    {
+    	nop(); /* trap */
```

Comment:
why do we have these nop's and not while(1)'s that will get stuck here? If this fails it will be hard for the customer to understand unless they are debugging?

Reply:
I do not have context on why this is a nop. This code is originally provided by an external source, I have just updated it to be compatible with the latest APIs of the OTA library. This looks like a completely new file because we removed the OTA library completely before adding in the new version.

I'm unsure if there is some kind of requirement for this to be here, the comment "trap" is kind of cryptic. The only thing that I can think of is as you mentioned, it's used for debugging so that you can set a break point there.

Are you convinced that the behavior that we want to take is a while(1) here? I agree that there isn't a great way to raise an error in this code, but I think that failing to write to flash is a possible runtime error that may not appear right away. I don't think that we should make the code get stuck here. If the device does fail to write to the flash, this will still be detected when the device goes to verify the signature of the image and fails.

I'm more in favor of adding a print statement here in addition to the nop (in case logging is disabled?). What do you think?

Reply:
What this looks like is that nop's re placed to enable you to place a breakpoint as a trap and stop the code at this location.

This is a common "trick" on embedded systems.

My concern is that this is an OTA PAL which purports to be re-usable by people for their projects on this platform, and if you are using that and an error happens you surely do not want it just ignored so this seems to be a real problem. 

If this PAL is dangerous/not ready to use we should probably be removing it.

Sounds like this is an old problem though and you are now having to deal with it. Let me ask around what we should be doing here.

Reply:
Ok it looks like most of those traps are just for debug purposes and the failures would just make the OTA fail gracefully later so they are not critical. Since we did not change this I will just accept the PR for now, we will log an internal ticket to look at this a little more closely to ensure we did not miss something here during the review.